File: 0062-unique-paths/0062-unique-paths.py
from collections import deque
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        
        #m rows
        #n cols

        dp = {}

        def dfs(row, col):
            
            #Base Case [Get to the end [m - 1] [n- 1]]
            if row == m - 1 and col == n - 1:
                return 1
            
            if (row, col) in dp:
                return dp[(row, col)]
            
            #Bounds
            if row < 0 or row == m or col < 0 or col == n:
                return 0
            
            #Call dfs
            dp[(row, col)] = dfs(row + 1, col) + dfs(row, col + 1)

            return dp[(row, col)]

        return dfs(0, 0)





        

File: 0062-unique-paths/0062-unique-paths.cpp
using namespace std;

class Solution {
public:
    int uniquePaths(int m, int n) {
        
        int rows = m;
        int cols = n;
        //Top-down - using DP TABLE
        vector<vector<int>> dp(rows, vector<int>(cols, 0));
        dp[0][0] = 1;

        //First row
        for (int i = 1; i < cols; i++)
        {
            dp[0][i] = 1;
        }

        //First col
        for (int j = 1; j < rows; j++)
        {
            dp[j][0] = 1;
        }

        //We want to loop everything in middle!!!
        for (int i = 1; i < rows; i++)
        {
            for (int j = 1; j < cols; j++)
            {
                dp[i][j] = dp[i - 1][j] + dp[i][j-1];
            }
        }

        return dp[rows - 1][cols - 1];
    }
};